src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const allStatement = [];
+
+  const toModified = { ...state };
+
+  for (const action of actions) {
+    change(toModified, action);
+    allStatement.push(Object.assign({}, toModified));
+  }
+
+  return allStatement;
+}
+
+function change(state, action) {
+  switch (action.type) {
+    case 'addProperties':
+      Object.assign(state, action.extraData);
+      break;
+
+    case 'removeProperties':
+      for (const key of action.keysToRemove) {
+        delete state[key];
+      }
+      break;
+
+    case 'clear':
+      for (const key in state) {
+        delete state[key];
+      }
+      break;
+  }
```

Comment:
1. right here you can just invoke a loop
2. at the end of the loop you can add changes
3. after loop you can just return an array
```suggestion
  for (const action of actions) {
    switch (action.type) {
      case 'addProperties':
        Object.assign(state, action.extraData);
        break;
  
      case 'removeProperties':
        for (const key of action.keysToRemove) {
          delete state[key];
        }
        break;
  
      case 'clear':
        for (const key in state) {
          delete state[key];
        }
        break;
    }
    
    allStatement.push({ ...coppiedState });
  }
  
  return allStatement;
```